examples/text-classification/run_glue.py

```diff
@@ -353,20 +391,29 @@ def compute_metrics(p: EvalPrediction):
         train_result = trainer.train(
             model_path=model_args.model_name_or_path if os.path.isdir(model_args.model_name_or_path) else None
         )
-        metrics = train_result.metrics
-
-        trainer.save_model()  # Saves the tokenizer too for easy upload
+        ##### $custom ####
+        if is_run_on_sagemaker():
+            trainer.save_model(os.environ["SM_MODEL_DIR"])  # Saves the tokenizer too for easy upload
+        else:
+            trainer.save_model()  # Saves the tokenizer too for easy upload
+        ####  $custom end ####
 
         output_train_file = os.path.join(training_args.output_dir, "train_results.txt")
         if trainer.is_world_process_zero():
             with open(output_train_file, "w") as writer:
                 logger.info("***** Train results *****")
-                for key, value in sorted(metrics.items()):
+                for key, value in sorted(train_result._asdict().items()):
                     logger.info(f"  {key} = {value}")
                     writer.write(f"{key} = {value}\n")
 
-            # Need to save the state, since Trainer.save_model saves only the tokenizer with the model
-            trainer.state.save_to_json(os.path.join(training_args.output_dir, "trainer_state.json"))
+            ##### $custom ####
+            if is_run_on_sagemaker():
+                # Need to save the state, since Trainer.save_model saves only the tokenizer with the model
+                trainer.state.save_to_json(os.path.join(os.environ["SM_MODEL_DIR"], "trainer_state.json"))
+            else:
+                # Need to save the state, since Trainer.save_model saves only the tokenizer with the model
+                trainer.state.save_to_json(os.path.join(training_args.output_dir, "trainer_state.json"))
+            ####  $custom end ####
```

Comment:
this `if` clause should not be needed if we tweak `.output_dir` above, right?

Reply:
Yes. But we have to be careful since we have in Sagemaker 2 Output directories. One for logs and files (e.g. `eval_result.txt`) and the other is for the model and associated files (config etc). 

`SM_MODEL_DIR` represents the `model` directory and `SM_OUTPUT_DATA_DIR` represents the directory for logs and files.

Reply:
Yes, that we can probably add "upstream" in TrainingArguments if we don't already have it (the possibility of having different output dirs for model and logs)